src/libasr/pass/template_visitor.cpp

```diff
@@ -0,0 +1,298 @@
+                        }
+                        result = left_value >> right_value;
+                        break;
+                    }
+                    default: { LFORTRAN_ASSERT(false); } // should never happen
+                }
+                value = ASR::down_cast<ASR::expr_t>(ASR::make_IntegerConstant_t(al, loc, result, dest_type));
+            }
+            tmp = ASR::make_IntegerBinOp_t(al, loc, left, op, right, dest_type, value);
+        } else if (ASRUtils::is_real(*dest_type)) {
+            if (op == ASR::binopType::BitAnd || op == ASR::binopType::BitOr || op == ASR::binopType::BitXor ||
+                op == ASR::binopType::BitLShift || op == ASR::binopType::BitRShift) {
+                throw SemanticError("Unsupported binary operation on floats: '" + ASRUtils::binop_to_str_python(op) + "'", loc);
+            }
+            right = cast_helper(left_type, right);
+            dest_type = ASRUtils::expr_type(right);
+            if (ASRUtils::expr_value(left) != nullptr && ASRUtils::expr_value(right) != nullptr) {
+                double left_value = ASR::down_cast<ASR::RealConstant_t>(ASRUtils::expr_value(left))->m_r;
+                double right_value = ASR::down_cast<ASR::RealConstant_t>(ASRUtils::expr_value(right))->m_r;
+                double result;
+                switch (op) {
+                    case (ASR::binopType::Add): { result = left_value + right_value; break; }
+                    case (ASR::binopType::Sub): { result = left_value - right_value; break; }
+                    case (ASR::binopType::Mul): { result = left_value * right_value; break; }
+                    case (ASR::binopType::Div): { result = left_value / right_value; break; }
+                    case (ASR::binopType::Pow): { result = std::pow(left_value, right_value); break; }
+                    default: { LFORTRAN_ASSERT(false); }
+                }
+                value = ASR::down_cast<ASR::expr_t>(ASR::make_RealConstant_t(al, loc, result, dest_type));
+            }
+            tmp = ASR::make_RealBinOp_t(al, loc, left, op, right, dest_type, value);
+        }        
+    }
+
+    ASR::expr_t *cast_helper(ASR::ttype_t *left_type, ASR::expr_t *right,
+            bool is_assign=false) {
+        ASR::ttype_t *right_type = ASRUtils::type_get_past_pointer(ASRUtils::expr_type(right));
+        if (ASRUtils::is_integer(*left_type) && ASRUtils::is_integer(*right_type)) {
+            int lkind = ASR::down_cast<ASR::Integer_t>(left_type)->m_kind;
+            int rkind = ASR::down_cast<ASR::Integer_t>(right_type)->m_kind;
+            if ((is_assign && (lkind != rkind)) || (lkind > rkind)) {
+                return ASR::down_cast<ASR::expr_t>(ASRUtils::make_Cast_t_value(
+                    al, right->base.loc, right, ASR::cast_kindType::IntegerToInteger,
+                    left_type));
+            }
+        }
+        return right;
+    }
+
+};
```

Comment:
I think this is same as AST to ASR conversion logic for BinOp nodes. This is inevitable because at the end of the day templates are nothing but at AST level (no semantics available because no concrete type information is available). Probably this would lead to repeating the whole AST to ASR logic for all the `TemplateXYZ` nodes. May be we can make common utility functions for this purpose which can be called in AST to ASR transition and here in this pass as well.

Reply:
I was a bit bothered by this part as well, because both `make_BinOp_helper` and `cast_helper` were defined within the AST visitor, it wasn't a function available outside of the class.